src/tls/server.c

```diff
@@ -0,0 +1,902 @@
+handle_accept (void)
+{
+  int fd;
+  Connection *con;
+  struct epoll_event ev = { .events = EPOLLIN };
+
+  /* accept and create new connection */
+  fd = accept4 (server.listen_fd, NULL, NULL, SOCK_CLOEXEC);
+  if (fd < 0)
+    {
+      if (errno == EINTR)
+        return;
+      err (1, "failed to accept connection");
+    }
+  con = connection_new (fd);
+
+  /* epoll the connection fd */
+  ev.data.fd = fd;
+  if (epoll_ctl (server.epollfd, EPOLL_CTL_ADD, fd, &ev) < 0)
+    err (1, "Failed to epoll connection fd");
+
+  /* add to our Connections list */
+  block_sigchld (true);
+  con->next = server.connections;
+  server.connections = con;
+  block_sigchld (false);
+}
+
+/**
+ * handle_connection_data_first: Handle first event on client fd
+ *
+ * Check the very first byte of a new connection to tell apart TLS from plain
+ * HTTP. Initialize TLS and the ws instance.
+ */
+static void
+handle_connection_data_first (int fd, Connection *con)
+{
+  char b;
+  int ret;
+
+  assert (!con->ws);
+
+  /* peek the first byte and see if it's a TLS connection (starting with 22).
+     We can assume that there is some data to read, as this is called in response
+     to an epoll event. */
+  ret = recv (fd, &b, 1, MSG_PEEK);
+  if (ret < 0)
+    err (1, "failed to peek first byte");
+  if (ret == 0) /* EOF */
+    {
+      debug ("client disconnected without sending any data");
+      remove_connection (fd, NULL);
+      return;
+    }
+
+  if (b == 22)
+    {
+      gnutls_session_t session;
+
+      debug ("first byte is %i, initializing TLS", (int) b);
+
+      if (!server.x509_cred)
+        {
+          warnx ("got TLS connection, but our server does not have a certificate/key; refusing");
+          remove_connection (fd, NULL);
+          return;
+        }
+
+      gnutls_check (gnutls_init (&session, GNUTLS_SERVER));
+      gnutls_check (gnutls_priority_set (session, server.priority_cache));
+      gnutls_check (gnutls_credentials_set (session, GNUTLS_CRD_CERTIFICATE, server.x509_cred));
+
+      gnutls_certificate_server_set_request (
+          session,
+          (server.client_cert_mode == CERT_REQUEST) ? GNUTLS_CERT_REQUEST : GNUTLS_CERT_IGNORE);
+      gnutls_handshake_set_timeout (session, GNUTLS_DEFAULT_HANDSHAKE_TIMEOUT);
+
+      gnutls_transport_set_int (session, fd);
+
+      TLS_RETRY_BLOCK (ret, gnutls_handshake (session));
+      if (ret < 0)
+        {
+          warnx ("TLS handshake failed: %s", gnutls_strerror (ret));
+          remove_connection (fd, NULL);
+          return;
+        }
+
+      debug ("TLS handshake completed");
+
+      connection_set_tls_session (con, session);
+    }
+
+  connection_init_ws (con);
+  if (!con->ws)
+    remove_connection (fd, NULL);
+}
+
+static int
+send_all (int socket, const char *data, size_t data_size, int flags)
+{
+  int r;
+
+  while (data_size > 0)
+  {
+    r = send (socket, data, data_size, flags);
+
+    if (r < 0)
+      {
+        if (errno == EAGAIN || errno == EINTR)
```

Comment:
EAGAIN means that the syscall would have blocked.  Retrying the syscall right away without select/epoll will result in a tight loop until progress can be made.